paperqa/agents/main.py

```diff
@@ -333,19 +338,25 @@ def __init__(
         on_env_step_callback: (
             Callable[[list[Message], float, bool, bool], Awaitable] | None
         ) = None,
+        settings: Settings | None = None,
     ):
         self.env = env
         self.on_env_reset_callback = on_env_reset_callback
         self.on_agent_action_callback = on_agent_action_callback
         self.on_env_step_callback = on_env_step_callback
+        self.settings = settings
 
     async def after_agent_get_asv(self, traj_id: str, *args) -> None:  # noqa: ARG002
         if self.on_agent_action_callback is not None:
             await self.on_agent_action_callback(*args)
 
-    async def after_env_reset(self, traj_id: str, *_) -> None:  # noqa: ARG002
+    async def after_env_reset(
+        self, traj_id: str, obs: list[Message], tools: list[Tool]  # noqa: ARG002
+    ) -> None:
         if self.on_env_reset_callback is not None:
             await self.on_env_reset_callback(self.env.state)
+        if self.settings:
+            self.settings.adapt_tools(tools)
```

Comment:
Is it ok to have settings in this class?

Also, there is something weird in this implementation: This will run `adapt_tools` every time an `after_env_reset` callback runs. I'm not sure if I like it.

The problem with passing the `adapt_tools` function as a callback (which I think would make more sense) is that the `after_env_reset` callback is defined to only receive the state.

Thoughts on this?

Reply:
> The problem with passing the `adapt_tools` function as a callback (which I think would make more sense) is that the `after_env_reset` callback is defined to only receive the state.

I am not following, `after_env_reset` receives `obs`, `tools`, `traj_id`, etc. We don't pass state at all?

> Is it ok to have settings in this class?

It's a breaking change in the signature, now seeing this it may be cleaner just to make a second focused callback:

```python
class LDPAdjustToolsForAgentCallback(Callback):
    def __init__(self, settings: Settings):
        self._settings = settings

    async def after_env_reset(
        self, traj_id: str, obs: list[Message], tools: list[Tool]
    ) -> None:
        self._settings.adjust_tools_for_agent_llm(tools)
```